r-package/grf/R/merge_forests.R

```diff
@@ -0,0 +1,78 @@
+#' Merges a list of forests that were grown using the same data into one large forest.
+#'
+#' @param forest_list A `list` of forests to be concatenated. 
+#'                        All forests must be of the same type, and the type must be a subclass of `grf`.
+#'                        
+#' @param compute.oob.predictions Whether OOB predictions on training set should be precomputed.
+#'        Note that even if OOB predictions have already been precomputed for the forests in 'forest_list',
+#'        those predictions are not used. Instead, a new set of oob predictions is computed anew using the 
+#'        larger forest.
+#' 
+#' @return A single forest containing all the trees in each forest in the input list.
+#'
+#' @examples \dontrun{
+#' # Train standard regression forests
+#' n = 50; p = 10
+#' X = matrix(rnorm(n*p), n, p)
+#' Y = X[,1] * rnorm(n)
+#' r.forest1 = regression_forest(X, Y, compute.oob.predictions = FALSE, num.trees = 100)
+#' r.forest2 = regression_forest(X, Y, compute.oob.predictions = FALSE, num.trees = 100)
+#'
+#' # Join the forests together. The resulting forest will contain 200 trees.
+#' big_rf = merge_forests(list(r.forest1, r.forest2))
+#' }
+#' 
+#' @export
+merge_forests <- function(forest_list, compute.oob.predictions=TRUE) {
+  
+  validate_forest_list(forest_list)
+  
+  first_forest <- forest_list[[1]]
+  data <- create_data_matrices(first_forest$X.orig, first_forest$Y.orig)
+  
+  big_forest <- merge(forest_list, data$default, data$sparse)
+  
+  big_forest[["ci.group.size"]] <- first_forest$ci.group.size
+  big_forest[["X.orig"]] <- first_forest$X.orig
+  big_forest[["Y.orig"]] <- first_forest$Y.orig
+  big_forest[["clusters"]] <- first_forest$clusters
+  big_forest[["min.node.size"]] <- first_forest$min.node.size
+  
+  class(big_forest) <- class(first_forest)
+  
+  if (compute.oob.predictions) {
+    oob.pred <- predict(big_forest)
+    big_forest[["predictions"]] <- oob.pred$predictions
+    # Must include checks here because some big_forest types may have this 
+    # method not yet implemented
+    if (!is.null(big_forest["debiased.error"])) {
```

Comment:
Could you explain a little more about this check?

Reply:
I had in mind forests that output only "predictions", but not the other statistics. 

However, it seems that we don't really need that, or at least that this isn't the fix we need. For example, what should we do about `quantile_forests`?

```
X <- matrix(runif(1000), 500, 2)
Y <- matrix(runif(500))
rf <- grf::quantile_forest(X, Y)
pred <- predict(rf) 
class(pred)  # "matrix"
pred$predictions # Error: $ operator is invalid for atomic vectors 
```

Reply:
That seems like a bug to me, since the prediction objects should have the same structure across the different forest types (quantile forests should still contain the `predictions` subfield, and just omit `debiased.error` etc.). I would vote to not worry about this in the current PR, and plan to fix the discrepancy in predictions formats in a follow-up PR.